Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,380 @@
+         => ByteString -> Maybe (a, ByteString)
+_readInt bs = case uncons bs of
+    Just (w, rest)
+        | toDigit w <= 9 -> convertPos $ readWord64 bs   -- leading digit
+        | w == 0x2d      -> convertNeg $ readWord64 rest -- minus sign
+        | w == 0x2b      -> convertPos $ readWord64 rest -- plus sign
+    _                    -> Nothing                      -- not a number
+  where
+    -- Conversion from Word64 to a positive Int is value-preserving provided
+    -- the input does not exceed @maxBound :: Int@ as a 64-bit value.
+    convertPos (Just (w64, str))
+        | w64 <= imax64 @a
+             = let !i = fromIntegral @Word64 @a w64
+                in Just (i, str)
+    convertPos _ = Nothing
+    -- We accept '-0' as well as '+0'.  Conversion from Word64 to a negated Int
+    -- is value-preserving provided the input does not exceed the absolute
+    -- value of @'minBound' :: 'Int'@.
+    convertNeg (Just (w64, str))
+        | w64 <= imin64 @a
+             = let !i = negate $ fromIntegral @Word64 @a w64
+                in Just (i, str)
+    convertNeg _ = Nothing
+{-# SPECIALISE _readInt :: ByteString -> Maybe (Int, ByteString) #-}
+{-# SPECIALISE _readInt :: ByteString -> Maybe (Int8, ByteString) #-}
+{-# SPECIALISE _readInt :: ByteString -> Maybe (Int16, ByteString) #-}
+{-# SPECIALISE _readInt :: ByteString -> Maybe (Int32, ByteString) #-}
+{-# SPECIALISE _readInt :: ByteString -> Maybe (Int64, ByteString) #-}
+
+----- Word functions
+
+-- | Try to read a 'Word' value from the 'ByteString', returning
+-- @Just (val, str)@ on success, where @val@ is the value read and @str@ is the
+-- rest of the input string.  If the sequence of digits decodes to a value
+-- larger than can be represented by a 'Word', the returned value will be
+-- 'Nothing'.
+--
+-- 'readWord' does not ignore leading whitespace, the value must start with a
+-- decimal digit immediately at the beginning of the input string.  Leading @+@
+-- signs are not accepted.
+--
+-- ==== __Examples__
+-- >>> readWord "1729 sum of cubes"
+-- Just (1729," sum of cubes")
+-- >>> readWord "+1729 has an explicit sign"
+-- Nothing
+-- >>> readWord "not a decimal number")
+-- Nothing
+-- >>> readWord "12345678901234567890 overflows maxBound")
+-- Nothing
+--
+readWord :: ByteString -> Maybe (Word, ByteString)
+{-# INLINE readWord #-}
+readWord = _readWord @Word
+
+-- | A variant of 'readWord' specialised to 'Word64'.
+readWord64 :: ByteString -> Maybe (Word64, ByteString)
+#ifdef BYTESTRING_STRICT
+readWord64 (BI.BS fp len) = BI.accursedUnutterablePerformIO $
+    BI.unsafeWithForeignPtr fp $ \ptr -> do
+        let end = ptr `plusPtr` len
+        (!n, !a, !inRange) <- digits end ptr 0 0
+        if inRange
+            then if n < len
+                 then let rest = BI.BS (fp `BI.plusForeignPtr` n) (len - n)
+                       in return $! result n a rest
+                 else return $! result n a empty
```

Comment:
Surely `drop` or `unsafeDrop` is cleaner?

Reply:
It occurs to me now that `unsafeDrop` would give some different (perhaps undesirable) behavior with respect to finalizers in the empty-result case. So perhaps this is best left as-is, even if it is a bit ugly.

Reply:
In the empty result case `unsafeDrop` retains a reference to the whole `ByteString`, returning an empty length fragment. As a minor optimisation, I chose to return `empty`.  It is perhaps not uncommon to call readInt and friends on buffers that contain only the number...